include/trial/protocol/json/partial/copy_and_look_ahead.hpp

```diff
@@ -0,0 +1,176 @@
+#ifndef TRIAL_PROTOCOL_JSON_PARTIAL_COPY_AND_LOOK_AHEAD_HPP
+#define TRIAL_PROTOCOL_JSON_PARTIAL_COPY_AND_LOOK_AHEAD_HPP
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Copyright (C) 2021 Vinícius dos Santos Oliveira <vini.ipsmaker@gmail.com>
+//
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          http://www.boost.org/LICENSE_1_0.txt)
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include <trial/protocol/json/partial/skip.hpp>
+
+namespace trial
+{
+namespace protocol
+{
+namespace json
+{
+namespace partial
+{
+
+namespace detail {
+template<class CharT>
+struct CopyAndLookAheadCollector
+{
+    using view_type = typename basic_reader<CharT>::view_type;
+    using size_type = typename basic_reader<CharT>::size_type;
+
+    CopyAndLookAheadCollector(view_type key)
+        : pattern{key}
+    {}
+
+    void push_back(CharT v)
+    {
+        if (!process) return;
+
+        if (!pattern.starts_with(v)) {
+            process = false;
+            return;
+        }
+
+        pattern.remove_prefix(1);
+    }
+
+    void append(const CharT *data, size_type size)
+    {
+        if (!process) return;
+
+        view_type run{data, size};
+        if (!pattern.starts_with(run)) {
+            process = false;
+            return;
+        }
+
+        pattern.remove_prefix(size);
+    }
+
+    bool current_value_matches_pattern()
+    {
+        return process && pattern.size() == 0;
+    }
+
+    bool process = true;
+    view_type pattern;
+};
+} // namespace detail
+
+template<class CharT, class F>
+auto/*=bool*/ copy_and_look_ahead(const basic_reader<CharT> &reader_,
+                                  typename basic_reader<CharT>::view_type key,
+                                  F&& on_value, std::error_code &ec)
+    -> decltype(on_value(const_cast<basic_reader<CharT>&>(reader_)), true)
+{
+    assert(!ec);
+
+    if (reader_.symbol() != json::token::symbol::begin_object) {
+        ec = errc::field_not_found;
+        return false;
+    }
+
+    basic_reader<CharT> reader(reader_);
+
+    if (!reader.next()) {
+        ec = reader.error();
+        return false;
+    }
+
+    for (;;) {
+        if (reader.symbol() == json::token::symbol::end_object) {
+            ec = errc::field_not_found;
```

Comment:
Same `unexpected_token` comment as above.

Reply:
Actually this case is different. Here we ran over the whole object and didn't find the target key. A different error code — I recycled the one created for the `scanf()` PR — will allow the user to discern between an error in the JSON stream (was the JSON stream corrupt?) and errors on high-level algorithms applied on top of such stream (does the JSON DOM validates against the expected data schema?).

Reply:
Ok